src/m365/planner/commands/task/task-reference-add.ts

```diff
@@ -0,0 +1,112 @@
+import { Logger } from '../../../../cli';
+import { CommandOption } from '../../../../Command';
+import GlobalOptions from '../../../../GlobalOptions';
+import request from '../../../../request';
+import GraphCommand from '../../../base/GraphCommand';
+import commands from '../../commands';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  taskId: string;
+  url: string;
+  alias?: string;
+  type?: string;
+}
+
+class PlannerTaskReferenceAddCommand extends GraphCommand {  
+  public get name(): string {
+    return commands.TASK_REFERENCE_ADD;
+  }
+
+  public get description(): string {
+    return 'Adds a new reference to a Planner task';
+  }
+
+  public commandAction(logger: Logger, args: CommandArgs, cb: () => void): void {
+    this
+      .getTaskDetailsEtag(args.options.taskId)
+      .then(etag => {
+        const requestOptionsTaskDetails: any = {
+          url: `${this.resource}/v1.0/planner/tasks/${args.options.taskId}/details`,
+          headers: {
+            'accept': 'application/json;odata.metadata=none',
+            'If-Match': etag,
+            'Prefer': 'return=representation'
+          },
+          responseType: 'json',
+          data: {
+            references: {
+              [this.customURIEncoder(args.options.url)]: {
+                '@odata.type': 'microsoft.graph.plannerExternalReference',
+                previewPriority: ' !',
+                ...(args.options.alias && {alias: args.options.alias}),
+                ...(args.options.type && {type: args.options.type})
+              }
+            }
+          }
+        };
+
+        return request.patch(requestOptionsTaskDetails);
+      })
+      .then((res: any): void => {
+        logger.log(res.references);
+        cb();
+      }, (err: any): void => this.handleRejectedODataJsonPromise(err, logger, cb));
+  }
+
+  private getTaskDetailsEtag(taskId: string): Promise<string> {
+    const requestOptions: any = {
+      url: `${this.resource}/v1.0/planner/tasks/${encodeURIComponent(taskId)}/details`,
+      headers: {
+        accept: 'application/json'
+      },
+      responseType: 'json'
+    };
+
+    return request
+      .get(requestOptions)
+      .then((response: any) => {
+        const etag: string | undefined = response ? response['@odata.etag'] : undefined;
+
+        if (!etag) {
+          return Promise.reject(`Error fetching task details`);
+        }
+
+        return Promise.resolve(etag);
+      });
+  }
+
+  private customURIEncoder(value: string): string {
```

Comment:
Good idea! I'd also suggest that we change the name to a verb as it's not a class (thing) that we're instantiating but rather performing an action.

Reply:
`urlUtils` won't be the right location for this. Mainly because it's an encoder for property names in Open Types. So this doesn't directly relate to urls.

Reply:
In that case, it should go to formatting utils.